host/sgx/create.c

```diff
@@ -562,6 +572,104 @@ oe_result_t oe_sgx_validate_enclave_properties(
     return result;
 }
 
+#ifdef OE_WITH_EXPERIMENTAL_EEID
+static oe_result_t _eeid_resign(
+    oe_sgx_load_context_t* context,
+    oe_sgx_enclave_properties_t* properties)
+{
+    oe_result_t result = OE_OK;
+    oe_eeid_t* eeid = context->eeid;
+
+    if (eeid && eeid->data_size > 0)
+    {
+        sgx_sigstruct_t* sigstruct = (sgx_sigstruct_t*)properties->sigstruct;
+
+        OE_SHA256 ext_mrenclave;
+        oe_sha256_final(&context->hash_context, &ext_mrenclave);
+
+        OE_CHECK(oe_sgx_sign_enclave(
+            &ext_mrenclave,
+            properties->config.attributes,
+            properties->config.product_id,
+            properties->config.security_version,
+            OE_DEBUG_SIGN_KEY,
+            OE_DEBUG_SIGN_KEY_SIZE,
+            sigstruct));
+    }
+
+done:
+    return result;
+}
+
+static oe_result_t _add_eeid_pages(
+    oe_sgx_load_context_t* context,
+    oe_sgx_enclave_properties_t* props,
+    uint64_t enclave_addr,
+    uint64_t* vaddr,
+    uint64_t entry_point)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_eeid_t* eeid = context->eeid;
+
+    if (eeid)
+    {
+        if (props->header.size_settings.num_heap_pages != 0 ||
+            props->header.size_settings.num_stack_pages != 0 ||
+            props->header.size_settings.num_tcs != 1)
+            OE_RAISE(OE_INVALID_PARAMETER); // For now.
+
+        eeid->version = 1;
+        oe_sha256_context_t* hctx = &context->hash_context;
+        oe_sha256_save(hctx, eeid->hash_state.H, eeid->hash_state.N);
+        eeid->vaddr = *vaddr;
+        eeid->entry_point = entry_point;
+        eeid->signature_size = sizeof(sgx_sigstruct_t); // SGX only for now
+
+        size_t num_bytes = 8 + sizeof(oe_eeid_t) + eeid->data_size;
+        size_t num_pages =
+            num_bytes / OE_PAGE_SIZE + ((num_bytes % OE_PAGE_SIZE) ? 1 : 0);
+
+        sgx_sigstruct_t* sigstruct = (sgx_sigstruct_t*)&props->sigstruct;
+        memcpy(eeid->signature, (uint8_t*)sigstruct, sizeof(sgx_sigstruct_t));
+
+        oe_page_t* pages =
+            oe_memalign(OE_PAGE_SIZE, sizeof(oe_page_t) * num_pages);
+        *((uint64_t*)pages->data) = OE_EEID_MAGIC;
+        memcpy(
+            pages->data + sizeof(uint64_t),
+            eeid,
+            sizeof(oe_eeid_t) + eeid->data_size);
+
+        for (size_t i = 0; i < num_pages; i++)
+        {
+            uint64_t addr = enclave_addr + *vaddr;
+            uint64_t src = (uint64_t)(pages + i * OE_PAGE_SIZE);
+            uint64_t flags = SGX_SECINFO_REG | SGX_SECINFO_R | SGX_SECINFO_W;
+
+            OE_CHECK(oe_sgx_load_enclave_data(
+                context, enclave_addr, addr, src, flags, false));
```

Comment:
Why isn't the eeid data being measured into the extended MRENCLAVE value? The design indicated that it should:

> - measure the additional page containing the oe_eeid_t (filled with the entire oe_eeid_t, including data),

Reply:
It is; just a couple lines earlier it is copied onto the pages: `memcpy(pages->data + sizeof(uint64_t), eeid, ...);`

Reply:
@wintersteiger  I don't follow. Calling `oe_sgx_load_enclave_data` with false as the last parameter means it won't be measured.

Reply:
Darn! You're right, the flag is wrong even though the data is right!

Reply:
(Fixed properly now.)